### PR TITLE
8253385: annotation processors remove varargs information from record components

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this fix for a corner case bug. Essentially for code like:

```
import java.lang.annotation.*;

@Target({ElementType.TYPE_USE})
@interface Simple {}

record R(@Simple int ...val) {
    static void test() {
        R rec = new R(10, 20);
    }
}
```

assuming that there is an AP present, after the first annotation processing round, the record component's symbol is set to `null` and thus the flags info is lost. So once the symbols and their types are determined again after every annotation processing round, there is no recollection that the record component was a varargs. Remember that the AST for the record component declaration is not kept but it is used to extract information from it and to derive compiler generated code like fields, canonical constructor, etc. The proposed solution is to keep the information about the vararg-ness of the record component in a dedicated field in the record component so that this information can be recovered later on as needed.

TIA,
Vicente

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ❌ (2/9 failed) | ❌ (1/9 failed) |

**Failed test tasks**
- [Windows x64 (hs/tier1 gc)](https://github.com/vicente-romero-oracle/jdk/runs/1354692029)
- [Windows x64 (hs/tier1 serviceability)](https://github.com/vicente-romero-oracle/jdk/runs/1354692079)
- [macOS x64 (hs/tier1 gc)](https://github.com/vicente-romero-oracle/jdk/runs/1354619662)

### Issue
 * [JDK-8253385](https://bugs.openjdk.java.net/browse/JDK-8253385): annotation processors remove varargs information from record components

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 379ee616a36589d2d30d5dd2e642bd349d4af75d

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/709/head:pull/709`
`$ git checkout pull/709`

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java [hunk 1]
```diff
@@ -1785,19 +1785,27 @@ public void setData(Object data) {
          */
         private final int pos;
 
+        private boolean isVarargs;
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:1788]: field can be final?

File: test/langtools/tools/javac/records/RecordCompilationTests.java [hunk 2]
```diff
@@ -1682,4 +1682,39 @@ record R() implements java.io.Serializable {}
...
+
+    public void testAnnotationsOnVarargsRecComp() {
+        assertOK(
+                """
+                import java.lang.annotation.*;
+
+                @Target({ElementType.TYPE_USE})
+                @interface Simple {}
+
+                record R(@Simple int ...val) {
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/records/RecordCompilationTests.java:1694]: (minor) unusual typography for "int ...val". Suggest "int... val"

File: test/langtools/tools/javac/records/RecordCompilationTests.java [hunk 3]
```diff
@@ -1682,4 +1682,39 @@ record R() implements java.io.Serializable {}
...
+                import java.lang.annotation.*;
+
+                @Target({ElementType.TYPE_USE})
+                @interface SimpleContainer{ Simple[] value(); }
+
+                @Repeatable(SimpleContainer.class)
+                @Target({ElementType.TYPE_USE})
+                @interface Simple {}
+
+                record R(@Simple int ...val) {
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/records/RecordCompilationTests.java:1712]: (minor) unusual typography for "int ...val". Suggest "int... val"

**Review by jonathan-gibbons — APPROVED** [review]
minor suggestions in comments

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:1788]: sure will do

**Comment by vicente-romero-oracle** [comment]
thanks for the review

**Pull request closed without merging**